### PR TITLE
feat: added support for batch request feature to WS server (#2461)

### LINKED ISSUE #2461: Cannot send batch requests via WS endpoints
### Description

Expected:
Should be able to send Batch requests via WS endpoint
Whats Happening:
Currently batch requests are not respected & no response returned

### Steps to reproduce

 [{
    "jsonrpc": "2.0",
    "method": "eth_blockNumber",
    "params": [],
    "id": 2
}
]

### Additional context

_No response_

### Hedera network

testnet

### Version

0.47.0-rc3

### Operating system

None

### PR DESCRIPTION
**Pull request by quiet-node**
**Description**:
- adds support for batch request feature to WS server
- adds WS neccessary config variables to configurations.md
- added two new predefined errors, WS_BATCH_REQUESTS_DISABLED and WS_BATCH_REQUESTS_AMOUNT_MAX_EXCEEDED, to JsonRpcError object
- added e2e tests to cover the new feature

Todo: https://github.com/hashgraph/hedera-json-rpc-relay/issues/2474

Fixes #2461

**Result**

- wscat:
image

- postman
image

image

**Notes for reviewer**:

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by AlfredoG87 — DISMISSED** [review]
LG. just a few nits! 👍

File: docs/configuration.md [hunk 1]
```diff
@@ -24,21 +24,23 @@ These properties are noted below and should be custom set per deployment.
...
+| ---------------------------- | ---------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `BATCH_REQUESTS_ENABLED`     | "false"          | Flag to disable or enable batch requests.                                                                                                                                                                               |
+| `BATCH_REQUESTS_MAX_SIZE`    | "100"            | Maximum number of requests allowed in a batch.                                                                                                                                                                          |
+| `WS_BATCH_REQUESTS_ENABLED`  | "false"          | Flag to enable batch requests on the websocket server.                                                                                                                                                                  |
```
[the remark this answers is not in the document]
Comment (quiet-node) [reply docs/configuration.md:31]: Oh it was mentioned in the design doc https://github.com/hashgraph/hedera-json-rpc-relay/blob/main/docs/design/batch-request.md#configuration.

File: packages/ws-server/src/webSocketServer.ts [hunk 2]
```diff
@@ -113,84 +112,78 @@ app.ws.use(async (ctx) => {
...
-    // Validate request's params
-    try {
-      const methodValidations = Validator.METHODS[method];
-
-      if (methodValidations) {
-        Validator.validateParams(params, methodValidations);
+      // send error if batch request feature is not enabled
+      if (!getWsBatchRequestsEnabled()) {
+        const batchRequestDisabledError = predefined.WS_BATCH_REQUESTS_DISABLED;
+        logger.error(`${connectionIdPrefix} ${requestIdPrefix}: ${JSON.stringify(batchRequestDisabledError)}`);
```
[the remark this answers is not in the document]
Comment (quiet-node) [reply packages/ws-server/src/webSocketServer.ts:122]: Copy that!

File: packages/ws-server/src/webSocketServer.ts [hunk 3]
```diff
@@ -113,84 +112,78 @@ app.ws.use(async (ctx) => {
...
-            connectionIdPrefix,
-          );
-          response = jsonResp(request.id, null, relayResult);
+      // send error if batch request exceed max batch size
+      if (request.length > getBatchRequestsMaxSize()) {
+        const batchRequestAmountMaxExceed = predefined.BATCH_REQUESTS_AMOUNT_MAX_EXCEEDED(
+          request.length,
+          getBatchRequestsMaxSize(),
+        );
+        logger.error(`${connectionIdPrefix} ${requestIdPrefix}: ${JSON.stringify(batchRequestAmountMaxExceed)}`);
```
[the remark this answers is not in the document]
Comment (quiet-node) [reply packages/ws-server/src/webSocketServer.ts:133]: Heard!

File: packages/ws-server/src/webSocketServer.ts [hunk 4]
```diff
@@ -113,84 +112,78 @@ app.ws.use(async (ctx) => {
...
+      sendToClient(ctx.websocket, request, response, logger, requestIdPrefix, connectionIdPrefix);
     }
 
     // Calculate the duration of the connection
     const msgEndTime = process.hrtime(msgStartTime);
     const msgDurationInMiliSeconds = (msgEndTime[0] + msgEndTime[1] / 1e9) * 1000; // Convert duration to miliseconds
 
     // Update the connection duration histogram with the calculated duration
-    wsMetricRegistry.getHistogram('messageDuration').labels(method).observe(msgDurationInMiliSeconds);
+    const methodLable = Array.isArray(request) ? WS_CONSTANTS.BATCH_REQUEST_METHOD_NAME : request.method;
```
[the remark this answers is not in the document]
Comment (quiet-node) [reply packages/ws-server/src/webSocketServer.ts:185]: ohh lool good catch! That's so funny `methodLable`

File: docs/configuration.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlfredoG87) [reply docs/configuration.md:31]: Yeah, we launched on http server with feature disabled as default, but right now is enabled by default.
we should probably do the same, maybe we can pilot test it for 1 sprint and if everything is good, we can set it as true by default next sprint.

please open up an issue so we don't forget.

**Review by AlfredoG87 — APPROVED** [review]
LGTM

**Review by AlfredoG87 — APPROVED** [review]

File: docs/configuration.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (quiet-node) [reply docs/configuration.md]: Gotcha I'll update the configs. [Ticket](https://github.com/hashgraph/hedera-json-rpc-relay/issues/2477) created

**Review by AlfredoG87 — APPROVED** [review]
LGTM

**Pull request merged**